### PR TITLE
Fixed <output> inside repeat has absolute path instead of relative

### PR DESCRIPTION
**Pull request by gushil**
#446

### CODE REVIEW

**Comment by MartijnR** [comment]
@ukanga, just pinging to check if you're able to look at this Ona-reported issue. :)

File: pyxform/survey.py [hunk 1]
```diff
@@ -722,34 +729,40 @@ def itext(self):
...
-
-                        if media_type == "guidance":
-                            itext_nodes.append(
-                                node(
-                                    "value",
-                                    value,
-                                    form="guidance",
-                                    toParseString=output_inserted,
-                                )
+                    if label_type == "hint" or media_type == "long":
```
Comment (lognaturel) [inline pyxform/survey.py:732]: Why this test?

File: pyxform/survey.py [hunk 2]
```diff
@@ -603,7 +603,14 @@ def _setup_choice_translations(name, choice_value, itext_id):
                     self._translations[d["lang"]][d["path"]] = self._translations[
                         d["lang"]
                     ].get(d["path"], {})
-                    self._translations[d["lang"]][d["path"]].update({"long": d["text"]})
+                    if "element" in d:
+                        self._translations[d["lang"]][d["path"]].update(
+                            {"long": {"text": d["text"], "element": d["element"]}}
```
Comment (lognaturel) [inline pyxform/survey.py:608]: Making the context part of the `long` form doesn't make sense to me (and also it doesn't work for guidance hints). Could it be a sibling to the form? It seems it applies to the lang/path combination and any form should use it.

**Comment by MartijnR** [comment]
Thanks for your review @lognaturel! Just wanted to let you know OpenClinica has slightly delayed deploying the new pyxform version (they use a very old version), and I've heard @gushil may pick this up in a few weeks.

File: pyxform/survey.py [hunk 3]
```diff
@@ -603,7 +603,14 @@ def _setup_choice_translations(name, choice_value, itext_id):
                     self._translations[d["lang"]][d["path"]] = self._translations[
                         d["lang"]
                     ].get(d["path"], {})
-                    self._translations[d["lang"]][d["path"]].update({"long": d["text"]})
+                    if "element" in d:
```
[the remark this answers is not in the document]
Comment (gushil) [reply pyxform/survey.py]: This is for checking whether there is an "element" key in d. the "element" is used for storing a reference to the current translated element that will be used for a context in the insert_output_values method.

File: pyxform/survey.py (hunk 2 above)
Comment (gushil) [reply pyxform/survey.py]: If you don't mind, any suggestion about it? Changing dictionary key is doable. I'll change that with the next commit.

File: pyxform/survey.py (hunk 1 above)
Comment (gushil) [reply pyxform/survey.py]: This is just a little code refactor. Will be changed in the next commit.

File: pyxform/tests_v1/test_repeat.py [hunk 4]
```diff
@@ -266,3 +266,22 @@ def test_hints_are_present_within_groups(self):
     </group>"""  # noqa
 
         self.assertPyxformXform(md=md, xml__contains=[expected], run_odk_validate=True)
+
+    def test_inside_repeat_has_relative_path(self):
```
[the remark this answers is not in the document]
Comment (gushil) [reply pyxform/tests_v1/test_repeat.py]: Will do. Thanks for the suggestion.

File: pyxform/survey.py (hunk 1 above)
Comment (lognaturel) [reply pyxform/survey.py]: Whoops, sorry about that!

File: pyxform/survey.py (hunk 2 above)
Comment (lognaturel) [reply pyxform/survey.py]: Unfortunately what I had in mind would require some bigger changes to `itext`. Let's go with what you have.

**Comment by lognaturel** [comment]
@MartijnR I don't think there's anything in the spec that specifically addresses `output` in a `text` translation for labels in secondary instances. It seems wacky but people use it in lieu of #38: https://forum.getodk.org/t/xlsform-index-repeat-calculate-filter-choice-odk-collect-v1-27-crash/27411. Is that supposed to be supported? If so, this PR should generate relative refs for those.

**Comment by MartijnR** [comment]
@lognaturel, you're referring to the output below, right? (using the test.xlsx form from that forum thread)

```xml
 <instance id="hm_name">
                <root>
                    <item>
                        <filter>1</filter>
                        <label>${name1}</label><!-- uh oh -->
                        <name>1</name>
                    </item>
```

I tried to figure out a syntax below. Not sure if I succeeded in that, and whether this implementable in pyxform. It wouldn't work in Enketo in the near future. It also wouldn't work for external instances, which to me is a fair argument to not support it all.

It is extremely convoluted xlsform (and xform) syntax indeed. I agree that https://github.com/XLSForm/pyxform/issues/38 would seem so much more sensible.

```xml
<itext>
      <text id="a1">
             <value><output value="../name1"></value>
      </text>
</itext>

 <instance id="hm_name">
                <root>
                    <item>
                        <filter>1</filter>
                        <itextId>a1</itextId>
                        <name>1</name>
                    </item>

...

            <itemset nodeset="instance('hm_name')/root/item[filter &lt; ( /data/qb_1 +1)]">
                <value ref="name" />
                <label ref="itext(itextId)" />
            </itemset>
</select>

```

**Comment by MartijnR** [comment]
Oh LOL https://github.com/XLSForm/pyxform/issues/403

**Comment by lognaturel** [comment]
Ahhh! I knew I'd seen something fly by related to this but couldn't think of where. Let's not worry about it here and instead try to get #38 which would be really great. So I think this is very close to being ready.

**Comment by pbowen-oc** [comment]
@lognaturel @MartijnR - I can't tell from the discussion above - will this update remove the existing behavior that allowed item references in choice labels? We have used this to very good effect (never in a translated form configuration). It is very helpful for allowing the user to (for example) associate a medication and an adverse event or associate a procedure with a previously identified tumor.

In our experience, the only restriction we found was that we needed some text in the label other than the item reference itself.

**Comment by lognaturel** [comment]
Can you please share a minimal XLSForm on #403 that illustrates the pattern you want to use, @pbowen-oc? This PR doesn’t change existing behavior around this so let’s continue the conversation there.

**Comment by pbowen-oc** [comment]
@lognaturel - Thanks, I have added an example file to #403

File: pyxform/tests_v1/test_repeat.py [hunk 5]
```diff
@@ -187,6 +187,25 @@ def test_indexed_repeat_relative_path(self):
             ],
         )
 
+    def test_output_with_translation_relative_path(self):
+        md = """
+        | survey |              |             |                |              |
+        |        | type         | name        | label::English | calculation  |
+        |        | begin repeat | member      |                |              |
+        |        | calculate    | pos         |                | position(..) |
+        |        | text         | member_name | Name of ${pos} |              |
```
[the remark this answers is not in the document]
Comment (gushil) [reply pyxform/tests_v1/test_repeat.py:196]: Ok, will add this test.

File: pyxform/survey.py [hunk 6]
```diff
@@ -591,19 +591,31 @@ def _setup_choice_translations(name, choice_value, itext_id):
...
-                    self._translations[d["lang"]][hint_path].update(
-                        {"guidance": d["text"]}
+                    translation_path = d["path"].replace("guidance_hint", "hint")
+                    translation_key = "guidance"
+
+                self._translations[d["lang"]][translation_path] = self._translations[
+                    d["lang"]
+                ].get(translation_path, {})
+
+                if "output_context" in d:
```
[the remark this answers is not in the document]
Comment (gushil) [reply pyxform/survey.py:606]: It looks like some old tests (not in tests_v1) is not passed if I remove this test. However, I've found the cause of those errors and will be fixed by next commit.

File: pyxform/survey.py [hunk 7]
```diff
@@ -591,19 +591,31 @@ def _setup_choice_translations(name, choice_value, itext_id):
...
+                                "output_context": d["output_context"],
+                            }
+                        }
                     )
                 else:
-                    self._translations[d["lang"]][d["path"]] = self._translations[
-                        d["lang"]
-                    ].get(d["path"], {})
-                    self._translations[d["lang"]][d["path"]].update({"long": d["text"]})
+                    self._translations[d["lang"]][translation_path].update(
```
[the remark this answers is not in the document]
Comment (gushil) [reply pyxform/survey.py:616]: Ok, will remove this.

File: pyxform/survey.py [hunk 8]
```diff
@@ -722,12 +734,18 @@ def itext(self):
...
 
                 for media_type, media_value in content.items():
-                    # There is a odk/jr bug where hints can't have a value
-                    # for the "form" attribute.
-                    # This is my workaround.
-                    if label_type == "hint":
+                    if (
+                        isinstance(media_value, dict)
+                        and "text" in media_value
+                        and "output_context" in media_value
```
[the remark this answers is not in the document]
Comment (gushil) [reply pyxform/survey.py:740]: Got it.

File: pyxform/survey.py [hunk 9]
```diff
@@ -591,19 +591,31 @@ def _setup_choice_translations(name, choice_value, itext_id):
         self._translations = defaultdict(dict)  # pylint: disable=W0201
         for element in self.iter_descendants():
             for d in element.get_translations(self.default_language):
+
+                translation_path = d["path"]
+                translation_key = "long"
```
[the remark this answers is not in the document]
Comment (gushil) [reply pyxform/survey.py:596]: It looks like I have to make changes to one other place if I changed translation_key. You'll see in my next commit.

File: pyxform/survey.py [hunk 10]
```diff
@@ -722,12 +734,18 @@ def itext(self):
...
+                    if (
+                        isinstance(media_value, dict)
+                        and "text" in media_value
+                        and "output_context" in media_value
+                    ):
+                        value, output_inserted = self.insert_output_values(
+                            media_value["text"], context=media_value["output_context"]
+                        )
+                    else:
                         value, output_inserted = self.insert_output_values(media_value)
```
[the remark this answers is not in the document]
Comment (gushil) [reply pyxform/survey.py:746]: Ok

File: pyxform/survey.py [hunk 11]
```diff
@@ -743,15 +752,13 @@ def itext(self):
                             )
                         continue
 
-                    if media_type == "long":
-                        value, output_inserted = self.insert_output_values(media_value)
+                    if media_type == "form":
```
Comment (gushil) [inline pyxform/survey.py:755]: I have to make changes to this line if I changed `translation_key` to `form`.

File: pyxform/survey.py [hunk 12]
```diff
@@ -593,7 +593,7 @@ def _setup_choice_translations(name, choice_value, itext_id):
             for d in element.get_translations(self.default_language):
 
                 translation_path = d["path"]
-                translation_key = "long"
+                translation_key = "form"
```
Comment (lognaturel) [inline pyxform/survey.py:596]: I meant changing the name of the variable to `form`. `long` and `guidance` are values for the form attribute.

File: pyxform/survey_element.py [hunk 13]
```diff
@@ -271,6 +271,7 @@ def get_translations(self, default_language):
                 for lang, text in constraint_msg.items():
                     yield {
                         "path": self._translation_path("jr:constraintMsg"),
+                        "output_context": self,
```
Comment (lognaturel) [inline pyxform/survey_element.py:274]: Oh, great! There was a bug before because all of these need relative paths as well.

**Comment by lognaturel** [comment]
I took the liberty of renaming the variable that represents the translation form and adding tests for the additional translated columns. Unfortunately it looks like I forgot to run black and I had to quickly leave my computer so things are in a weird state. I’ll fix and merge this evening. 

Thanks for doing this and for taking my feedback, @gushil!

**Pull request merged**